### ISSUE TITLE
Create "covidcastify" functions to import other data sources and use them alongside COVIDcast indicators

### ISSUE THREAD

**Issue by capnrefsmmat** [cmu-delphi/covidcast#322]
Suppose a user finds a new data source online. It has, say, a bunch of data for specific counties on various days, measuring some signal of interest related to COVID.

The user would like to load this into the covidcast package to make maps, correlate it with our other signals, maybe put it into a forecasting pipeline, etc.

We should have a flexible function that lets the user pass a data frame, specify which columns are dates and which are FIPS codes, name the source and signal(s), and process this to return a data frame compatible with all the other plotting, mapping, and correlation functions:

```r
foo <- read.csv("some-random-dataset.csv")
foo_c <- covidcastify(foo, value = "pcr_test_num", time_value = "date", time_format = "%m/%d/%Y", ...)
```

Then we should make a vignette that takes a third-party dataset -- say, COVID Tracking Project data, or another popular source -- and loads it into the package for analysis.

Things to think about:

* How to handle issue dates. Do we make our functions that use issue dates (e.g. `covidcast_cor()` calls `latest_issue()`) ignore them if no `issue` column is present, or do we make every "covidcastified" data frame have an `issue` column?
* Are there other columns that functions expect to exist, but that may not be present in outside data? How do we ensure those columns either are present or are gracefully ignored?
* How to attach metadata.
* How to flexibly handle any kind of outside input data in weird formats.

**Comment by ryantibs** [comment]
Thanks @capnrefsmmat for creating this issue.  I think you've summarized it all nicely and generally agree with everything.

Just wanted to raise one more point: I don't think we should *require* a user to call `covidcastify()` in order to use all our functions (plotting, correlations, etc.).  It would be just an important to rewrite the functions so that they can be called on a straight-up data frame with as few columns as possible.  

Example: for plotting, all you should need is `geo_value`, `time_value`, and `value`, and the rest can be inferred on-the-fly (likely, the `geo_type` can be inferred from the `geo_value`, and the range can be dynamically set---just as it is now, when the metadata mean and standard deviation are missing).

**Comment by capnrefsmmat** [comment]
I see `covidcastify()` as a design decision to keep our code simple: either `covidcastify()` handles augmenting the user's data frame with a minimum set of requirements, or every one of our other functions must grow extra complications to handle different cases, and we also need to write unit tests for all those cases. So we have to test `latest_issue` with both `covidcast_signal` objects and `data.frame` objects, and so on for all our other internal functions. I expect that could get pretty convoluted pretty quickly.

There might be a way to get around that, though. We can try looking at a few functions and see just how hard it is and what modifications would be required.

(My inspiration here is [sf](https://cran.r-project.org/package=sf), which can take input of many different types and turn it into a special spatial data frame, with a geography column and extra attributes that all of sf's machinery can then use correctly. It's not a huge burden to have to turn things into sf objects first.)

**Comment by ryantibs** [comment]
Actually I think a simple way of achieving what I'm asking for, without a major testing headache, is just to put a call to `covidcastify()` at the beginning of each function's body.  After all, `covidcastify()` should be idempotent, and should be very fast (unless we've implemented it poorly).  So there will be no hurt, and we can keep tests simple.  

And best of all, it will prevent the user from having to explicitly call `covidcastify()`: we can make this function public and explain its purpose, but if somebody doesn't want to or forgets to call it, they can still use our tools.  See any problems or have concerns?

**Comment by capnrefsmmat** [comment]
Ah, that could work. And, to be idiomatic, we could call it `as.covidcast`, and even have `as.covidcast.data.frame` and `as.covidcast.ts` or whatever other objects we might want covidcast to be able to deal with.

That doesn't translate to the Python package, but there's an analogy to `np.asarray` here.

**Comment by chinandrew** [comment]
Do we have any opinions on what `source` and `signal` should be for an arbitrary data source? We may be able to use the input variable name as one of those fields, and then having the user specify the other one (and default to empty string)

**Comment by capnrefsmmat** [comment]
I'm working on a draft of this feature for the R client, as a function `as.covidcast_signal` with S3 methods for different data types. I rely on the user to specify a signal name, since using the input variable name works until a user does something like

```r
as.covidcast_signal(foo %>% filter(date >= "2020-10-01"), ...)
```

and you get a funky long signal name.

Draft PR coming soon.

**Comment by capnrefsmmat** [comment]
Released in 0.4.0.
